### PR TITLE
Multiple cleanups and fix for #355

### LINKED ISSUE #355: "Should have been resolved after parsing"
I'm getting a "Should have been resolved after parsing" message from `Type::canonical_type()`. I printed more detail about the type:
```
thread 'main' panicked at 'Should have been resolved after parsing! Type { name: Some("category_type"), layout: Some(Layout { size: 1, align: 1, packed: false }), kind: ResolvedTypeRef(ItemId(43009)), is_const: false }', /home/engshare/third-party2/rust-bindgen/HEAD/src/build-centos6-native/build/libbindgen/src/ir/ty.rs:359
```

This appears to be https://github.com/facebook/folly/blob/master/folly/FBString.h#L598-L604:
```
  typedef uint8_t category_type;

  enum class Category : category_type {
    isSmall = 0,
    isMedium = kIsLittleEndian ? 0x80 : 0x2,
    isLarge = kIsLittleEndian ? 0x40 : 0x1,
  };
```

### PR DESCRIPTION
**Pull request by emilio**
Fixes #355 

Also improves the output of #354.

r? @fitzgen

### CODE REVIEW

**Comment by jsgf** [comment]
Confirm that this fixes the problem I reported in #355 in the full source.

File: libbindgen/src/codegen/mod.rs [hunk 1]
```diff
@@ -2010,14 +2010,10 @@ impl CodeGenerator for Function {
...
-        let signature = signature_item.kind().expect_type();
+        let signature = signature_item.kind().expect_type().canonical_type(ctx);
         let signature = match *signature.kind() {
             TypeKind::Function(ref sig) => sig,
-            TypeKind::ResolvedTypeRef(ref item_id) => {
-                let item = ctx.resolve_item(*item_id);
-                return self.codegen(ctx, result, _whitelisted_items, item);
-            },
-            _ => panic!("How?")
+            _ => panic!("How did this end up here {:?}?", signature)
```
Comment (fitzgen) [inline libbindgen/src/codegen/mod.rs]: Soft preference for a more descriptive message here:

> Signature's kind is not TypeKind::Function?! {:?}

File: libbindgen/src/ir/enum_ty.rs [hunk 2]
```diff
@@ -53,22 +53,18 @@ impl Enum {
...
         let mut variants = vec![];
 
-        let is_signed = match repr {
-            Some(repr) => {
-                let repr_type = ctx.resolve_type(repr);
-                match *repr_type.canonical_type(ctx).kind() {
+        // Assume signedness since the default type by the C standard is an int.
+        let is_signed = repr
+            .and_then(|r| ctx.resolve_type(r).safe_canonical_type(ctx))
+            .map_or(true, |ty| {
```
Comment (fitzgen) [inline libbindgen/src/ir/enum_ty.rs]: So much nicer :)

File: libbindgen/src/ir/int.rs [hunk 3]
```diff
@@ -91,6 +91,21 @@ impl IntKind {
         }
     }
 
+    /// If this type has a known size, return it (in bytes). This is to
+    /// alleviate libclang sometimes not giving us a layout (like in the case
+    /// when an enum is defined inside a class with template parameters).
+    pub fn known_size(&self) -> Option<usize> {
```
Comment (fitzgen) [inline libbindgen/src/ir/int.rs]: Great :)

**Comment by emilio** [comment]
Updated the error message.

@bors-servo r=fitzgen

**Pull request merged**